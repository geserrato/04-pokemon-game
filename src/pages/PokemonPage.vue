<template>
  <div v-if="hasPokemon">
    <h1>Quien es este Pok√©mon</h1>
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon"/>
    <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer($event)"/>

    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame">Nuevo Juego</button>
    </template>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import PokemonPicture from '@/components/PokemonPicture.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'

// Helpers
import getPokemonOptions from '@/helpers/getPokemonOptions.ts'

// Interface
import { IPokemon } from '@/interfaces/IPokemon'

export default defineComponent({
  name: 'PokemonPage',
  components: {
    PokemonPicture,
    PokemonOptions
  },
  data () {
    return {
      pokemonArr: [] as IPokemon[],
      pokemon: {} as IPokemon,
      hasPokemon: false,
      showPokemon: false,
      showAnswer: false,
      message: ''
    }
  },
  methods: {
    mixPokemonArray: async function () {
      this.hasPokemon = false
      this.pokemonArr = await getPokemonOptions()
      const rndInit = Math.floor(Math.random() * 4)
      this.pokemon = this.pokemonArr[rndInit]
      this.hasPokemon = true
    },
    checkAnswer (selectedId: number) {
      this.showPokemon = true
      this.showAnswer = true

      if (selectedId === this.pokemon.id) {
        this.message = `Si es, ${this.pokemon.name}`
      } else {
        this.message = `Noo, es ${this.pokemon.name}`
      }
    },
    newGame () {
      this.showPokemon = false
      this.showAnswer = false
      this.hasPokemon = false
      this.message = ''
      this.mixPokemonArray()
    }
  },
  mounted () {
    this.mixPokemonArray()
  }
})
</script>
