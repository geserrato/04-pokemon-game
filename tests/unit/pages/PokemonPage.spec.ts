import { shallowMount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage.vue'
import { IPokemon } from '@/interfaces/IPokemon'
import { pokemons } from '../mocks/pokemons.mock'

describe('PokemonPage', () => {
  let wrapper: any

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage)
  })

  test('debe de hacer macth con el snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('debe de llamar mixPokemonArray al montar el componente', () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
    shallowMount(PokemonPage)
    expect(mixPokemonArraySpy).toHaveBeenCalled()
  })

  test('debe de hacer match con el snapshot cuando carga los pokemon', () => {
    const wrapper = shallowMount(PokemonPage, {
      data () {
        return {
          pokemonArr: pokemons as IPokemon[],
          pokemon: pokemons[2] as IPokemon,
          hasPokemon: true,
          showPokemon: false,
          showAnswer: false,
          message: ''
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('debe de mostrar los componente de PokemonPicture y PokemonOptions', () => {
    const wrapper = shallowMount(PokemonPage, {
      data () {
        return {
          pokemonArr: pokemons as IPokemon[],
          pokemon: pokemons[2] as IPokemon,
          hasPokemon: true,
          showPokemon: false,
          showAnswer: false,
          message: ''
        }
      }
    })

    const PokemonPicture = wrapper.find('pokemon-picture-stub')
    const PokemonOptions = wrapper.find('pokemon-options-stub')

    expect(PokemonPicture.exists).toBeTruthy()
    expect(PokemonOptions.exists).toBeTruthy()
    expect(PokemonPicture.attributes('pokemonid')).toBe('3')
    expect(PokemonOptions.attributes('pokemons')).toBeTruthy()
  })

  test('pruebas en check answer', async () => {
    const wrapper = shallowMount(PokemonPage, {
      data () {
        return {
          pokemonArr: pokemons as IPokemon[],
          pokemon: pokemons[1] as IPokemon,
          hasPokemon: true,
          showPokemon: false,
          showAnswer: false,
          message: ''
        }
      }
    })

    await wrapper.vm.checkAnswer(2)
    expect(wrapper.find('h2').exists).toBeTruthy()
    expect(wrapper.vm.showPokemon).toBeTruthy()
    expect(wrapper.vm.showAnswer).toBeTruthy()
    expect(wrapper.vm.message).toBe(`Si es, ${pokemons[1].name}`)
  })
})
