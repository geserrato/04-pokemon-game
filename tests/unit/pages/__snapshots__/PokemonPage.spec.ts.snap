// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage debe de hacer macth con el snapshot 1`] = `<!--v-if-->`;

exports[`PokemonPage debe de hacer match con el snapshot cuando carga los pokemon 1`] = `
<div>
  <h1>Quien es este Pok√©mon</h1>
  <pokemon-picture-stub pokemonid="3" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
